src/main/java/at/ac/tuwien/kr/alpha/solver/TrailAssignment.java

```diff
@@ -0,0 +1,727 @@
+		outOfOrderLiterals = new ArrayList<>();
+		highestDecisionLevelContainingOutOfOrderLiterals = 0;
+		trail = new ArrayList<>();
+		trailIndicesOfDecisionLevels = new ArrayList<>();
+		trailIndicesOfDecisionLevels.add(0);
+		nextPositionInTrail = 0;
+		newAssignmentsIterator = 0;
+		newAssignmentsPositionInTrail = 0;
+		assignmentsForChoicePosition = 0;
+	}
+
+	@Override
+	public void registerCallbackOnChange(int atom) {
+		callbackUponChange[atom] = true;
+	}
+
+	@Override
+	public void setCallback(ChoiceManager choiceManager) {
+		choiceManagerCallback = choiceManager;
+	}
+
+	@Override
+	public boolean isAssigned(int atom) {
+		return values[atom] != 0;
+	}
+
+	private class TrailPollable implements Pollable {
+
+		@Override
+		public int peek() {
+			return trail.get(nextPositionInTrail);
+		}
+
+		@Override
+		public int remove() {
+			Integer current = peek();
+			nextPositionInTrail++;
+			return current;
+		}
+
+		@Override
+		public boolean isEmpty() {
+			return nextPositionInTrail >= trail.size();
+		}
+	}
+
+	@Override
+	public Pollable getAssignmentsToProcess() {
+		return new TrailPollable();
+	}
+
+	@Override
+	public int getRealWeakDecisionLevel(int atom) {
+		if (getTruth(atom) == null) {
+			return -1;
+		}
+		int lowestDecisionLevelForAtom = getWeakDecisionLevel(atom);
+		for (OutOfOrderLiteral outOfOrderLiteral : outOfOrderLiterals) {
+			if (outOfOrderLiteral.atom == atom && outOfOrderLiteral.decisionLevel < lowestDecisionLevelForAtom) {
+				lowestDecisionLevelForAtom = outOfOrderLiteral.decisionLevel;
+			}
+		}
+		return lowestDecisionLevelForAtom;
+	}
+
+	/**
+	 * Searches out-of-order literals for the lowest decision level the atom is assigned in.
+	 * @param atom the atom to check.
+	 * @return Integer.MAX_VALUE if the atom is not assigned out-of-order, otherwise the lowest decision level it is assigned.
+	 */
+	public int getOutOfOrderDecisionLevel(int atom) {
+		int lowestDecisionLevel = Integer.MAX_VALUE;
+		for (OutOfOrderLiteral outOfOrderLiteral : outOfOrderLiterals) {
+			if (outOfOrderLiteral.atom == atom && outOfOrderLiteral.decisionLevel < lowestDecisionLevel) {
+				lowestDecisionLevel = outOfOrderLiteral.decisionLevel;
+			}
+		}
+		return lowestDecisionLevel;
+	}
+
+	public int getOutOfOrderStrongDecisionLevel(int atom) {
+		int lowestDecisionLevel = Integer.MAX_VALUE;
+		for (OutOfOrderLiteral outOfOrderLiteral : outOfOrderLiterals) {
+			if (outOfOrderLiteral.atom == atom && outOfOrderLiteral.value == TRUE && outOfOrderLiteral.decisionLevel < lowestDecisionLevel) {
+				lowestDecisionLevel = outOfOrderLiteral.decisionLevel;
+			}
+		}
+		return lowestDecisionLevel;
+	}
+
+	private void informCallback(Integer atom) {
+		if (callbackUponChange[atom]) {
+			choiceManagerCallback.callbackOnChanged(atom);
+		}
+	}
+
+	private void removeLastDecisionLevel() {
+		// Remove all atoms recorded in the highest decision level.
+		Integer start = trailIndicesOfDecisionLevels.get(getDecisionLevel());
```

Comment:
`int` could be used instead of `Integer` to make clear we are not expecting `null`